src/Utilities.cpp

```diff
@@ -283,125 +283,125 @@ double brentRoot(
+    // Setup for next iteration
+    evenOlderPoint = previousPoint;
+    previousPoint = currentPoint;
+    previousFunc = currentFunc;
+    nextFunc = func(nextPoint);
+    if (counterFunc * nextFunc < 0) {
+      currentPoint = nextPoint;
+      currentFunc = nextFunc;
+    }
+    else {
+      counterPoint = nextPoint;
+      counterFunc = nextFunc;
+    }
+  } while (++iteration < 30 && fabs(counterPoint - currentPoint) > epsilon);
 
-  double x0 = lowerBound;
-  double x1 = upperBound;
-  double f0 = func(x0);
-  double f1 = func(x1);
-  double diff = 0;
-  double x2 = 0;
-  double f2 = 0;
+  return nextPoint;
+}
 
-  // Make sure we bound the root (f = 0.0)
-  if (f0 * f1 > 0.0) {
-    throw std::invalid_argument("Function values at the boundaries have the same sign [secantRoot].");
+double evaluatePolynomial(
+  const std::vector<double> &coeffs,
+  double x)
+{
+  if (coeffs.empty()) {
+    throw std::invalid_argument("Polynomial coeffs must be non-empty.");
   }
-
-  // Order the bounds
-  if (f1 < f0) {
-    std::swap(x0, x1);
-    std::swap(f0, f1);
+  auto revIt = coeffs.crbegin();
+  double value = *revIt;
+  ++revIt;
+  for (; revIt != coeffs.crend(); ++revIt) {
+    value *= x;
+    value += *revIt;
   }
+  return value;
+}
 
-  for (int iteration=0; iteration < maxIters; iteration++) {
-    x2 = x1 - f1 * (x1 - x0)/(f1 - f0);
-    f2 = func(x2);
+double evaluatePolynomialDerivative(
+  const std::vector<double> &coeffs,
+  double x)
+{
+  if (coeffs.empty()) {
+    throw std::invalid_argument("Polynomial coeffs must be non-empty.");
+  }
+  int i = coeffs.size() - 1;
+  double value = i * coeffs[i];
+  for (; i > 0; --i) {
+    value *= x;
+    value += i * coeffs[i];
+  }
+  return value;
+}
 
-    // Update the bounds for the next iteration
-    if (f2 < 0.0) {
-      diff = x1 - x2;
-      x1 = x2;
-      f1 = f2;
-    }
-    else {
-      diff = x0 - x2;
-      x0 = x2;
-      f0 = f2;
+double polynomialRoot(
+  const std::vector<double> &coeffs,
+  double guess,
+  double threshold,
+  int maxIterations)
+{
+  double root = guess;
+  double polyValue = evaluatePolynomial(coeffs, root);
+  double polyDeriv = 0.0;
+  for (int iteration = 0; iteration < maxIterations+1; iteration++) {
+    if (fabs(polyValue) < threshold) {
+      return root;
     }
-
-    // Check to see if we're done
-    if ((fabs(diff) <= epsilon) || (f2 == 0.0) ) {
-      found = true;
-      break;
+    polyDeriv = evaluatePolynomialDerivative(coeffs, root);
+    if (fabs(polyDeriv) < 1e-15) {
+      throw std::invalid_argument("Derivative at guess is too close to 0.");
     }
+    root -= polyValue / polyDeriv;
+    polyValue = evaluatePolynomial(coeffs, root);
   }
-
-  if (found) {
-   return x2;
-  }
-  else {
-    throw csm::Error(
-    csm::Error::UNKNOWN_ERROR,
-    "Could not find a root of the function using the secant method",
-    "secantRoot");
-  }
+  throw std::invalid_argument("Root finder did not converge in given number of iterations");
```

Comment:
advocate for adding a number of iterations to the error so that the iterations can be "upped" if desired.